lxd/instance/instance_utils.go

```diff
@@ -1315,3 +1316,51 @@ func getProjectVolumes(s *state.State, inst Instance) (volumes map[volKey]*db.St
 
 	return volumes, nil
 }
+
+// PlanAttachedVolumesForSnapshot determines the list of custom volumes to
+// snapshot for the provided instance and sets "volatile.attached_volumes".
+func PlanAttachedVolumesForSnapshot(s *state.State, inst Instance, diskVolumesMode string) (attached []*api.StorageVolume, err error) {
+	switch diskVolumesMode {
+	case api.DiskVolumesModeRoot, "":
+		// Root disk only, nothing to do.
+		return nil, nil
+	case api.DiskVolumesModeAllExclusive:
+	default:
+		return nil, errors.New("Invalid disk volumes mode")
+	}
+
+	// Get custom storage volumes in the instance's effective project.
+	volumes, err := getProjectVolumes(s, inst)
+	if err != nil {
+		return nil, err
+	}
+
+	// Build slice of attached volumes and map of volUUID -> new snapshot UUID.
+	attached = make([]*api.StorageVolume, 0, len(volumes))
+	uuids := make(map[string]string, len(volumes))
+	for _, dev := range inst.ExpandedDevices() {
+		v, ok := volumes[volKey{pool: dev["pool"], name: dev["source"]}]
+		// If the device's source volume doesn't exist or the device isn't sourced by the volume, skip it.
+		if !ok || !deviceSourcedByVolume(v.StorageVolume, dev) {
+			continue
+		}
+
+		attached = append(attached, &v.StorageVolume)
+
+		volUUID := v.Config["volatile.uuid"]
```

Comment:
this volUUID section looks weird, if volUUID is empty then it repeatedly writes a new UUID over the `uuids[""]` key?